Refactor Nexus chart

stevehipwell opened the pull request:
# Overview

This PR is intended to refactor the _Nexus 3_ chart closer to the idiomatic chart that you get when templating with _Helm_ and improve the automated _Nexus_ configuration. The patterns used for this will also be applied to the other charts in this repo (specifically _Jira_ and _Confluence_).

# Changes

This is a semver major change to `v3`.

## Added

- Image pull secret support
- Service account creation or selection
- Service account annotations
- Service annotations
- PVC annotations
- Role configuration supported

## Changed

- Better documentation of values
- Changed default service port to `8081` which matches the container port
- Changed the way the JVM heap is configured to align with the [Nexus 3 docs](https://help.sonatype.com/repomanager3/installation/system-requirements#SystemRequirements-Memory), use `env.jvmMinHeapSize` & `env.jvmMaxHeapSize`
- Renamed `securityContext` to `podSecurityContext` (`securityContext` now refers to the security context of the main container)
- Require `caCerts.enabled` to be set to `true` to use provided certs
- Configuration updated to use REST API where possible
- LDAP config values have changed
- Cleanup script has been updated to work correctly on _Nexus v3.25.0_
- Repo script has been updated to work correctly on _Nexus v3.25.0_

## Deprecated

- n/a

## Removed

- Support of RAM percentage in the JVM configuration (see changed section above)

Jasper-Ben commented:
Issues I encountered so far:
- LDAP not working, due to missing secret (should hopefully be fixed by the suggested changes above)
- cleanup policies still not showing up
- repo is not created (maybe due to missing cleanup policy)

[used values.yml](https://github.com/stevehipwell/helm-charts/files/5000058/values.txt)

Jasper-Ben commented:
LDAP works again after https://github.com/stevehipwell/helm-charts/pull/56/commits/a38630c9e9220921ca305bd211d9f75b9924be73 :+1:

stevehipwell commented:
@Jasper-Ben if LDAP config fails then all subsequent config fails, do you want to try again with the LDAP password changes?

stevehipwell commented:
@Jasper-Ben what about the cleanup now LDAP is working?

Jasper-Ben commented:
@stevehipwell unfortunately still nothing. :disappointed: same goes for the repo.

Jasper-Ben commented:
> @Jasper-Ben if LDAP config fails then all subsequent config fails, do you want to try again with the LDAP password changes?

Even when LDAP was broken, the configuration itself still went through (according to the logs). LDAP was also configured, just missing the bind password. These issues should therefore be unrelated.

Jasper-Ben commented:
The configuration succeeds according to the logs, even though they don't.

stevehipwell commented:
@Jasper-Ben are there no errors recorded in the logs?

Jasper-Ben commented:
@stevehipwell no, only the 404 that are to be expected, as the scripts do not exist before uploading:

```
Updating script /opt/sonatype/nexus/conf/repo.groovy.

2020-07-30 09:22:29,523+0000 WARN  [qtp1376681745-41] admin org.sonatype.nexus.siesta.internal.WebappExceptionMapper - (ID 8e7ed3f1-cc84-4569-8602-b8328631b13c) Response: [404] (no entity/body); mapped from: javax.ws.rs.NotFoundException: Script with name: 'repo' not found

Script /opt/sonatype/nexus/conf/repo.groovy updated.

Updating script /opt/sonatype/nexus/conf/task.groovy.

2020-07-30 09:22:29,547+0000 WARN  [qtp1376681745-42] admin org.sonatype.nexus.siesta.internal.WebappExceptionMapper - (ID 1414dcbd-1ba0-46c7-af5c-7740da8c0cc8) Response: [404] (no entity/body); mapped from: javax.ws.rs.NotFoundException: Script with name: 'task' not found

Script /opt/sonatype/nexus/conf/task.groovy updated.

Configuration run successfully!
```

stevehipwell commented:
@Jasper-Ben did you not get a 404 for the cleanup script?

stevehipwell commented:
@Jasper-Ben your cleanup and repo config values aren't indented correctly.

Jasper-Ben commented:
> @Jasper-Ben your cleanup and repo config values aren't indented correctly.

whoops! That indeed is a problem :sweat_smile:

Jasper-Ben commented on charts/nexus3/files/cleanup.groovy:
```diff
@@ -2,19 +2,24 @@ import org.sonatype.nexus.cleanup.storage.CleanupPolicy;
 } else {
-  def policy = policyStorage.newCleanupPolicy()
-  policyStorage.add(policy);
+  def newPolicy = cleanupPolicyStorage.newCleanupPolicy()
```
This is missing setting the name if the new cleanup policy

stevehipwell replied:
This should now be fixed!

Jasper-Ben commented on charts/nexus3/files/repo.groovy:
```diff
@@ -2,41 +2,36 @@ import groovy.json.JsonSlurper
+Configuration configuration
+if (existingRepository == null) {
+  configuration = repositoryManager.newConfiguration()
+  configuration.repositoryName = params.name,
```
delete `,` in lines 16-18

stevehipwell replied:
Done

Jasper-Ben commented:
works :+1:

Jasper-Ben commented on charts/nexus3/files/repo.groovy:
```diff
@@ -2,41 +2,36 @@ import groovy.json.JsonSlurper
-  repo.update(conf)
-  repo.start()
+  configuration.setOnline(params.online)
+  configuration.setAttributes(params.attributes)
```
This needs to be wrapped in a `if (params.attributes) ...`. Otherwise we cannot set the online status of an existing repository without changing its attributes.

stevehipwell replied:
It should work if you set the correct attributes for the repo you want to take offline.

Jasper-Ben replied:
@stevehipwell that is true, but this means I would have to "redefine" all the default repositories I just want to disable. So instead of
```
  repos:
    # disable default repositories
    - name: "maven-central"
      online: false
    - name: "maven-public"
      online: false
    - name: "maven-releases"
      online: false
    - name: "maven-snapshots"
      online: false
    - name: "nuget-group"
      online: false
    - name: "nuget-hosted"
      online: false
    - name: "nuget.org-proxy"
      online: false
```
this would become pretty messy (or at least require a `attributes: []` which I consider a bit hacky and I am not sure would work in the first place)

Jasper-Ben replied:
@stevehipwell I opened a pull request to address this

stevehipwell replied:
Thanks, I've merged this in.

stevehipwell commented:
@Jasper-Ben do you have anything else you want to add or are you waiting on this being released now?

Jasper-Ben commented:
@stevehipwell I am still trying to figure this out: https://github.com/stevehipwell/helm-charts/pull/62#issuecomment-668482640

I am not too familiar with the helm template language yet... maybe you have an idea?

Jasper-Ben commented:
the following works:

```
          {{- with .Values.extraVolumeMounts }}
            {{ toYaml . | nindent 12 }}
          {{- end }}
```

Jasper-Ben commented:
I opened a PR: https://github.com/stevehipwell/helm-charts/pull/63

Jasper-Ben commented:
as far as I am concerned this can be merged as well, once https://github.com/stevehipwell/helm-charts/pull/63 is included.

stevehipwell commented:
@Jasper-Ben I've rebased and squashed this PR in a way that your two contributions show up correctly. I'm about to merge this in for a v3.0.0 release.
